### ISSUE TITLE
discussion of Fake6502

### ISSUE THREAD

**Issue by gek169** [omarandlorraine/fake6502#8]
[redacted because we talked about it and i felt I was rude]

**Comment by omarandlorraine** [comment]
1. Wow, I can't believe I didn't notice that. Perhaps I introduced the bug before the test harness was put in place; I don't know.
I've put it right (I'm pretty sure!)

2. I don't know what the issue is there, maybe with your tests, maybe with my fork.  I might take a look in the coming weeks if I get time.

3. The choice of GPL is because I need this to be linked to my other GPL stuff.

4. Up yours, I do what I want

In general it is better if you've got 4 separate issue, to file four separate issues.

**Comment by gek169** [comment]
If I find more bugs i'll make more comments but i've been walking through the code i've adapted from Mike Chambers' original with a fine-toothed comb.

I don't mean to be an asshole when I make comments about your stylistic choices, i'm trying to tell you what issues I faced trying to integrate your code with other code. If I sound like an asshole, it's because I spent several hours changing `status` to `gContext.flags` only to find out that not only is your version completely broken but I have no idea how to fix it. Your friend's stochastic optimizer is undoubtedly completely bugged.

That said, I don't think moving to a struct is a bad decision. In fact, it's probably a really good decision, and I would like to migrate my fake6502 to a struct as well. 

My apologies if I sound like a jerk.

**Comment by gek169** [comment]
There is a difference between mike chamber's calculation of the overflow calculation and yours.
getValue's result is XOR'd with 0xff and the addition is made.
You use a native subtraction

However, value is used in the overflow calculation. So I suspect that your code will probably produce different overflow calculations than Mike Chambers' 

I am unsure which one is correct. Do you know?

Also, I wanted to ask you if there was some easy way I could take something like perfect6502 and brute force test every single instruction for identical behavior.

**Comment by omarandlorraine** [comment]
>If I sound like an asshole, it's because I spent several hours changing status to gContext.flags only to find out that not only is your version completely broken but I have no idea how to fix it.

I would recommend getting to grips with a decent editor. I'd suggest Vim. There's no way it should take several hours. And what does "completely broken" mean? The answer to that question will also point you in the right direction for fixing it. Or open another issue on here.

>That said, I don't think moving to a struct is a bad decision. In fact, it's probably a really good decision, and I would like to migrate my fake6502 to a struct as well.

It's the main reason I'm doing this. Also code deduplication which would be better for your instruction cache.

>My apologies if I sound like a jerk.

No trouble. Your feedback, although abrasively worded, has been valuable to me. I have reminded you of your motto from your page, "Let all that you do be done in love". 

>[overflow calculation] I am unsure which one is correct. Do you know?

I'm not sure. But I'd be tempted to trust Chambers on this one. Can you get a test-case in that exercises what you're talking about?

>Also, I wanted to ask you if there was some easy way I could take something like perfect6502 and brute force test every single instruction for identical behavior.

I have been thinking about the same thing. I think some of the trouble you'd run into doing every single instruction will be with the different versions of 6502, things like the jmp indirect bug, illegal instructions, especially the jam instructions etc etc. BUT you've given me the idea that my tests.c could easily be adapted to run on Mike Chambers' code, and other emulators as well probably. This could reveal more problems with my implementation and maybe others, by testing them against each other.

**Comment by gek169** [comment]
Alright, i'm glad you don't have beef with me. I thought you quoting my favorite quote was an indication that you were angry with me. It comes from Christ. I find it inspiring.

I am not angry with you at all and I respect your work. I think writing code for the public is a good thing. I personally avoid the GPL because I find that people are more appreciative of public domain code than GPL code, and my code gets used more. Despite not requiring an attribution requirement, I usually am credited somewhere. 

`It's the main reason I'm doing this. Also code deduplication which would be better for your instruction cache.`
Based on my experience writing instruction set simulators (I'm the original author of more than one. My most recent and expansive being SISA16, my own computer architecture with a working multitasking kernel and memory segmentation) this is only the second most important factor in speeding it up. (Note that a 6502 emulator needs not run at 1 billion IPS...)

The absolute best thing you can do is to have a separate dispatch for every single opcode. This does increase code size significantly, but the branch predictor has an easier time with it and you get a lower miss rate. My preferred way of doing this is with Computed Goto, a GnuC extension, however it is also possible to recreate by simply making a dispatch() macro with your switch case in it. While this causes the switch case to be duplicated in your binary and bloat the ISA simulator significantly (Neither GCC nor Clang seem smart enough to merge the jump tables...), it is still better than `while(1)switch(getop()){ case do_nop: break;}`

Lua employs the same technique in its virtual machine on compatible compilers. There was a very interesting cppcon talk that mentioned it briefly among other non-standard C/C++ extensions: https://youtu.be/IAdLwUXRUvg

This would mean, however, putting the entire emulator inside of a single function. I'm not a fan of that. Not to mention, it's a 6502. Those things only ever ran at most at 4-8 megahertz. I haven't run benchmarks, but just from playing around with kernalemu i can tell that fake6502 runs approximately two orders of magnitude faster than a real C64 (1 mhz iirc, 2 mhz clock with one tick for the cpu and one for the chipset) on my CPU. 

TLDR:
You are correct that smaller code size means fewer I-cache misses. However, it is hardly the only factor. Larger code can run faster due to quirks in how modern CPUs work.

That said, if you're benchmarking fake6502 on very old hardware (pre 2007) I think your theory would be entirely correct, and that branch prediction success rates would be far less important.

**Comment by gek169** [comment]
Again, I hope I don't mean to sound overly critical, but I have thought a little bit about the overall architecture of the emulator.

First of all, we need to consider the relevant use-cases. The most significant is obviously emulating 6502-based computer systems. The 6502 is quite slow and even a nintendo DS has no issues emulating an NES at full speed, so it's safe to say that use case is covered. It is also possible you could use the 6502 as an embeddable scripting language of sorts... There are compilers for it and everything. I'm sure nerds like the 8-bit guy and adrian black would love something like that. I have no idea if it'd be worth doing a JIT or dynarec.

So, I thought about how different architectural decisions might affect the emulator's usefulness as a scripting language.

1) Struct versus individual global variables.

Structs have a defined memory layout that, in my experience reading code, tends to be honored by the compiler. I don't know of any GNU extensions that let you put struct members into registers either, it's either the whole struct in a register, or none of it. You might be able to fit it in one of the SIMD registers...

It's possible the compiler may optimize away the struct layout if you use -fwhole-program and only use static variables,  but I find that unlikely. You might want to try checking what the codegen looks like on godbolt dot org. It might actually do it, I don't know.

It is possible, albeit perhaps a bad idea, to forcefully reserve a hardware register globally for the instruction set simulator in GCC. See: https://gcc.gnu.org/onlinedocs/gcc/Global-Register-Variables.html#Global-Reg-Vars

It would require careful programming, but I think it's how you would get the best speed without refactoring the entire emulator into a single function. It would also be possible to make the code work without the extension- simply use an ifdef to conditionally enable the register reservation. To guard the user's code, you simply have to have "stash" variables that are not registers, which you store the register variables into and restore from before executing instructions in the emulator. Implementing this would require renaming the variables in existing code wherever they occurred in the emulator, to use the static "register" versions, rather than the stashes (which would presumably keep the same names, to be compatible with kernalemu and tests.c)

2) Ease of programming

Obviously one factor we need to consider is how easy it is to drop-in the library.

No doubt the struct is superior in that regard. Object oriented programmers will feel right at home. Who the hell is going to know what a global "x" variable is? 

The struct, as you and webfra have noted, is also easier to make work with simulating multiple 6502s, which he uses for his stochastic optimizer (Very neat project btw! I commend you on that.)

3) Multithreading.

As the mike chambers' implementation stands, it would not work multi-threaded. Your version uses a struct so it might be possible to do it with yours.

If you wanted to simulate multiple 6502s on multiple hardware threads, you'd probably want some way for them to communicate with each other. My best idea would be to have a few atomic variables mapped into the 64k address space of each emulated processor, shared. 

You could go with something similar to the 6510 and treat address 0 as data direction and address 1 as the IO pins... and map those into atomic variables. Bits in $0 would select an atomic variable to read or write from, and reads or writes with address $1 would read or write the atomic variables. This would no doubt be the "most obvious" way to synchronize processors.

Alternatively, you could execute each 6502 in lock-step (in sets of so many clock cycles) and synchronize a page of memory in them on the 10000th or so clock cycle. Every byte in the page could be "owned" by a single processor, and thus any changes written to it would be reflected in the other processors, and any changes made by the other processors in bytes not owned by them would not be saved. You'd need to have some way of signalling to the emulated code that a synchronization page update occurred...

**Comment by gek169** [comment]
(off topic) in response to:
`I would recommend getting to grips with a decent editor. I'd suggest Vim. `
I used vim for about a year. Sorry, never going back to it. I use Micro. Vim is hard to use and it has weird keybinds. Whose idea was 'Yank' ?

I'd have the same issues with vim that I did while editing the code in Micro, in that the editor doesn't have an integrated way (that I know of) to change tokens according to C preprocessing rules. I could have used `cpp` (the c preprocessor) and literally used it to replace things in the code, but I wanted to preserve the macros that were already in the code. I'm aware some IDEs have a "variable renaming" feature, however i've had extremely bad experiences with IDEs in the past and I would rather spend 2-3 hours doing it by hand than discover a surprise "gift" that the IDE has left in my code without my knowledge. I would still like to do a variable renaming, so if you know an easy way to do it, please tell me because this a piece of refactoring work that frequently chews up massive amounts of my time.

Keep in mind that those 2-3 hours were spent editing all the references not only in fake6502 (<30 minutes) but also kernalemu (~2 hours) and your tests.c (30-40 minutes) 

Whenever I do a by-hand replacement, I have to use ctrl+f and scan through every occurrence of the text in the code, and verify that it is indeed the thing I want to replace and not something like a function's name which contains the variable name (exec6502 contains "x")

**Comment by omarandlorraine** [comment]
Not a real issue

**Comment by omarandlorraine** [comment]
The reason to speed the emulator up was also because of my superoptimiser. the bottleneck there is going to be how fast we can prove that 99% of the nonsense it generates is wrong, which depends on a really fast emulator. 

It seems you know more about how to get this running really fast, so I'd encourage you to give it a go! Then I might retire this fake6502 and use whatever you come up with instead.

**Comment by gek169** [comment]
Awfully generous of you!

**Comment by gek169** [comment]
I have just written a 65c02 implementation based on that by Paul Robson in the Commander X16 emulator and, not only have I found bugs in his versions of decimal mode, I have found bugs in your CMOS tests.

However, again, I'm not sure who's right here. I'm pretty sure that my decimal mode fixes are correct, but they don't match the X16 decimal mode (Probably because David Murray wrote x16's kernal in such a way that those decimal mode emualtion bugs never appeared. They are pathological cases...)

**Comment by omarandlorraine** [comment]
Can you open a new issue detailing which inputs make the decimal emulation bugs appear please

**Comment by gek169** [comment]
I'm still running my checks! I don't want to open an issue unless I know there really is an issue.

**Comment by gek169** [comment]
I'm not exactly a 6502 expert.
